File: revisiting_ctci/chapter4/RouteBetweenNodes_4_1.py
"""
Given a directed graph, design an algorithm to find out whther there is a 
route between two nodes
"""

from graph import Node, Graph


def IsRoute(TopNode, dest):
	""" conduct a breadth first search to see if there is a path between
		the start node and the destination"""
	nodes = [TopNode]
	while len(nodes) > 0 :
		current = nodes.pop(0)
		if current.Data == dest:
			return True
		nodes.extend(current.adj)

	return False


#conduct a breadth first search, beginning with a seed node and visiting
#all of its progeny, and their progeny in turn.




if __name__ == '__main__':

	#initiate a graph
	test_nodes = Graph()

	for i in range(6):
		x = i*3
		new = Node(x)

		if len(test_nodes.Nodes) > 0:
			test_nodes.Nodes[-1].Add(new)
		test_nodes.Add(new)


	IsRoute(test_nodes.Nodes[0], 14) #False

	IsRoute(test_nodes.Nodes[0], 12) #True

File: revisiting_ctci/chapter4/RouteBetweenNodes_4_1.go
/*
Given a directed graph, design an algorithm to find out whther there is a
route between two nodes
*/

package main

import (
	"./graph"
	"fmt"
)

// given a starting node, see if there is a route to a node
// containing the given value
func IsRoute(n graph.Node, v int) bool {
	nodes := []*graph.Node{&n}

	for len(nodes) > 0 {
		current_p := nodes[0]
		current := *current_p

		nodes = nodes[1:]

		if current.Data == v {
			return true
		}

		nodes = append(nodes, current.Adj...)
	}
	return false
}

func main() {

	n1 := graph.Node{Data: 0}
	n2 := graph.Node{Data: 5}
	n3 := graph.Node{Data: 10}
	n4 := graph.Node{Data: 15}
	n5 := graph.Node{Data: 20}

	n1.Add(&n2)
	n2.Add(&n3)
	n3.Add(&n4)
	n4.Add(&n5)

	g1 := graph.Graph{n1, n2, n3, n4, n5}

	start := g1[0]

	fmt.Println(start)
	fmt.Println(*(start.Adj[0]))

	fmt.Println("Graph being used:")
	fmt.Println(g1)

	fmt.Println("Path from 0 -> 19:")
	fmt.Println(IsRoute(start, 19))

	fmt.Println("Path from 0 -> 20:")
	fmt.Println(IsRoute(start, 20))
}
